bittide/src/Bittide/Wishbone.hs

```diff
@@ -551,18 +582,51 @@ timeWb ::
   , 1 <= DomainPeriod dom
   ) =>
   Circuit (Wishbone dom 'Standard addrW (Bytes 4)) ()
-timeWb = Circuit $ \(wbM2S, _) -> (mealy goMealy (0, 0) wbM2S, ())
+timeWb = Circuit $ \(wbM2S, _) -> (mealy goMealy (TimeGte, False, 0, 0) wbM2S, ())
  where
-  goMealy (frozen, count :: Unsigned 64) wbM2S = ((nextFrozen, succ count), wbS2M)
+  goMealy (cmp0, reqCmp0, scratch0 :: Unsigned 64, count :: Unsigned 64) wbM2S =
+    ((cmp1, reqCmp1, scratch1, succ count), wbS2M1)
    where
     freq = natToNum @(DomainToHz dom) :: Unsigned 64
-    nextFrozen = if isJust (head writes) then count else frozen
-    RegisterBank (splitAtI -> (frozenMsbs, frozenLsbs)) = getRegsBe @8 frozen
+
+    readVec :: Vec 7 (BitVector 32)
+    readVec =
+      0 -- Timer command is write-only, provide default 0 on read
+        :> (resize . pack $ cmp0)
+        :> pack scratchLsbs0
+        :> pack scratchMsbs0
+        :> pack freqLsbs
+        :> pack freqMsbs
+        :> (resize . pack $ runCmp)
+        :> Nil
+    (writes, wbS2M0) = wbToVec @4 @_ readVec wbM2S
+    (f0, f1, f2, f3, _f4, _f5, _f6) = vecToTuple writes
+
+    command :: Maybe TimeCmd
+    command = unpack . resize <$> f0
+    cmp1 = maybe cmp0 (unpack . resize) f1
+    cmdWaitForCmp = Just WaitForCmp == command
+    reqCmp1 = if reqCmp0 then not runCmp else cmdWaitForCmp
+
+    RegisterBank (splitAtI -> (scratchMsbs0, scratchLsbs0)) = getRegsBe @8 scratch0
+    scratch1 = case (bitCoerce f2, bitCoerce f3, command) of
+      (Just newLsbs, _, _) -> getDataBe $ RegisterBank (scratchMsbs0 ++ newLsbs)
+      (_, Just newMsbs, _) -> getDataBe $ RegisterBank (newMsbs ++ scratchLsbs0)
```

Comment:
Unfortunately here it's somewhat necessary, since `f2` and `f3` are `Maybe (BitVector 32)`, not `Maybe (Vec 4 (BitVector 8))`. They need to be the latter type since `RegisterBank` wants a `Vec 8 (BitVector 8)`, and `scratchMsbs0` and `scratchLsbs0` are providing the most/least significant halves of that respectively.

Reply:
Type conversion is indeed necessary, but instead of bitCoercing `Maybe (BitVector 32) -> Maybe (Vec 8 (BitVector 8))`, you can unpack `BitVector 32 -> Vec 8 (BitVector 8)` on the resulting lsbs and msbs respectively as per my suggestion.

Reply:
Oh, I see the difference now. I completely missed that.